#314: Resource shuffling and pinning

ShubertMunthali opened the pull request:
Issues #14 #15
Bounty #314

kirillt commented:
The "shuffle" button looks nice, I waited for this feature more than a year actually. Will be useful for learning languages.

UX comments:
- [ ] I described this poorly before, but we need quick `+` and `-` buttons in Gallery mode for increasing/decreasing score, arrows up/down also could be used as icons for these buttons. We still need action in "selection menu" (in Resources Grid screen), but it should be not "pin", but increase/decrease again and also "reset score" which erases selected resources' ids from scores storage.
- [ ] I realized that only now, that we need 2 states of "shuffle" button: "pressed" and "un-pressed". Default state is "un-pressed". When a user taps on it, it becomes "pressed" and resources are shuffled. When the user presses it second time, the button must become "un-pressed" and resources must become ordered again (according to chosen sorting). Otherwise, user would need work around this by changing tags in order to see ordered selection again.

kirillt commented:
@mdrlzy Could you guide Shubert regarding how to extract common logic of the storages into re-usable module/class? Also Shubert might need help/guidance with unit tests.

mdrlzy commented on app/src/main/java/space/taran/arknavigator/mvp/model/repo/scores/ScoreStorage.kt:
```diff
@@ -0,0 +1,17 @@
+    fun setScore(id: ResourceId, score: Score)
+
+    fun getScore(id: ResourceId): Score?
+
+    fun countScores(): Int
+
```
Why do we need this function?

ShubertMunthali replied:
It was being used to find a pinning position as a score but now that we have changed. It's no longer needed. It will be omitted

mdrlzy commented on app/src/main/java/space/taran/arknavigator/mvp/model/repo/scores/ScoreStorage.kt:
```diff
@@ -0,0 +1,17 @@
+    fun contains(id: ResourceId): Boolean
+
+    fun setScore(id: ResourceId, score: Score)
+
+    fun getScore(id: ResourceId): Score?
+
```
I think you can return non-nullable type since it can only be null in aggregated mode, but we have `contains` function for check

ShubertMunthali replied:
Yeah, and I realise it always returns a value

mdrlzy commented:
**Bug:** Pinned files stop being first if you change sorting

<details>
  <summary>Screenshot</summary>

![photo_2022-11-04_17-58-24](https://user-images.githubusercontent.com/85668474/199968243-5ef1490e-5a21-402b-ab69-2ed8cc2278ef.jpg)
</details>

mdrlzy commented:
Actually scores and pinned are two different things.
Looks like we need a separate storage for pinned ones.
@kirillt What do you think?

kirillt commented:
@mdrlzy yeah, pinning is different although looks nice. We could implement pinning using scores if we allow setting `MAX` score :) Both with pinning and scores, ordering must affect only positions of resource with the same scores.

In other words, first we sort by score (always when scores are enabled, we need a button or checkpoint for this), then we sort by the selected criteria.

kirillt commented on app/src/main/java/space/taran/arknavigator/mvp/model/repo/scores/PlainScoreStorage.kt:
```diff
@@ -53,23 +53,14 @@ class PlainScoreStorage(
         if (!scoreById.containsKey(id))
             error("Storage isn't aware of this resource id")
 
-        if (score <= 0) {
```
We allow negative scores indeed, but we should not store 0 scores since that could make the storage excessively big. `0` should be default value.

ShubertMunthali replied:
This is handled, no zero score is persisted

kirillt commented on app/src/main/java/space/taran/arknavigator/mvp/model/repo/scores/PlainScoreStorage.kt:
```diff
@@ -53,23 +53,14 @@ class PlainScoreStorage(
     }
 
-    override fun getScore(id: ResourceId) = scoreById[id]
+    override fun getScore(id: ResourceId) = scoreById[id]!!
```
Here we should fallback to `0` as the default value.

kirillt commented:
Star icon looks cool!

I think this feature would be really nice for UX:
- [ ] Button to hide scores, all its controls and do normal sorting.

I've found a tricky bug:
- [ ] If you open a folder. Edit some picture (overwrite it). Return back to Navigator, stay in Gallery mode, click `+` then `-`. First, the score goes to `1`, then to `-1`, after you can't make it any other number, only `1` or `-1`

mdrlzy commented on app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourcesGridPresenter.kt:
```diff
@@ -244,14 +310,24 @@ class ResourcesGridPresenter(
         }
     }
 
+    private fun sortUnpinnedResourcesOnly() {
+        val selection = this.selection.toSet()
+        this.selection = resources.filter {
+            selection.contains(it)
+        }.sortedWith(
+            compareBy {
+                !it.isPinned
+            }
+        )
+    }
+
```
I think it's better to rename it to `sortSelection`

mdrlzy reviewed (commented):
Looks nice

mdrlzy commented:
@kirillt I think it's a bad idea to use scores to determine pinned
For example, we define that `isPinned = score == Int.MAX_VALUE`
How to handle score increase/decrease in gallery? 
Do we need to prohibit changing score for pinned ones?

ShubertMunthali commented:
> @kirillt I think it's a bad idea to use scores to determine pinned For example, we define that `isPinned = score == Int.MAX_VALUE` How to handle score increase/decrease in gallery? Do we need to prohibit changing score for pinned ones?

I think it will be simpler to implement them separately if they can co-exist. I removed the pinning feature and added scores. We can add pinning later now.

kirillt commented on app/src/main/java/space/taran/arknavigator/utils/ComparatorUtils.kt:
```diff
@@ -16,3 +19,9 @@ fun <T> unequalCompareBy(comparator: Comparator<T>): Comparator<T> =
         } else result
     }
+
+fun compareByScores(scoreStorage: ScoreStorage) = Comparator<ResourceItem> { a, b ->
```
Better to move it somewhere closer to `ScoreStorage`, maybe even in the same module. Or to `ResourceGridPresenter` since it is used only there.

kirillt commented on app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourcesGridPresenter.kt:
```diff
@@ -316,13 +313,21 @@ class ResourcesGridPresenter(
+                        unequalCompareBy(
+                            if (sortByScores)
+                                compareResourcesByScores()
+                                    .then(comparator)
```
:+1:

mdrlzy commented on app/src/main/java/space/taran/arknavigator/mvp/presenter/GalleryPresenter.kt:
```diff
@@ -219,6 +236,21 @@ class GalleryPresenter(
+        withContext(Dispatchers.IO) {
+            scoreStorage.persist()
+        }
+        withContext(Dispatchers.Main) {
+            viewState.displayScore(score)
+        }
```
Here you also need to update scores on resources screen, for this you can use `viewState.notifyResourcesChanged()`

ShubertMunthali replied:
I have implemented it

mdrlzy commented on app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourcesGridPresenter.kt:
```diff
@@ -182,12 +206,72 @@ class ResourcesGridPresenter(
+        }
+    }
+
+    fun allowScoring() =
+        selectedResources.isNotEmpty() && selectedResources.size == 1 && sortByScores
```
Why do we disallow scoring for multiple resources?

ShubertMunthali replied:
I thought scoring should work for one resource every time. But that way, we can score many resources at once. I think that's ok

mdrlzy commented on app/src/main/java/space/taran/arknavigator/ui/fragments/ResourcesFragmentHelper.kt:
```diff
@@ -93,6 +94,26 @@ fun ResourcesFragment.setupAndShowSelectedResourcesMenu(menuBtn: View) {
                 .show(parentFragmentManager, null)
             popup.popupWindow.dismiss()
         }
+        with(btnIncreaseScore) {
+            isVisible = presenter.allowScoring()
+            setOnClickListener {
+                presenter.onIncreaseScoreClicked()
+                popup.popupWindow.dismiss()
+            }
+        }
+        with(btnDecreaseScore) {
+            isVisible = presenter.allowScoring()
+            setOnClickListener {
+                presenter.onDecreaseScoreClicked()
+                popup.popupWindow.dismiss()
+            }
+        }
```
I think popup can be left after changing score

ShubertMunthali replied:
Oh! yeah, sounds nice

kirillt commented on app/src/main/java/space/taran/arknavigator/ui/fragments/ResourcesFragment.kt:
```diff
@@ -87,6 +87,8 @@ class ResourcesFragment : MvpAppCompatFragment(), ResourcesView {
 
     private var tagsSelectorAdapter: TagsSelectorAdapter? = null
 
+    private var shuffledState = -1
```
Looks like it should be plain `Boolean` :)

ShubertMunthali replied:
Boolean brings conflicts when being updated with clicks on the button

ShubertMunthali replied:
I will find a better way to use it though :)

kirillt commented on app/src/main/java/space/taran/arknavigator/ui/fragments/ResourcesFragmentHelper.kt:
```diff
@@ -93,6 +94,45 @@ fun ResourcesFragment.setupAndShowSelectedResourcesMenu(menuBtn: View) {
+                    getString(
+                        R.string.resource_scores_erased,
+                        selectedSize,
+                        if (selectedSize == 1) "" else "s"
```
:+1:

ShubertMunthali replied:
Currently, we only have one language. If support for many languages is added we can always created the resources. I think for now that's simple.

kirillt replied:
Let's not mess with it for this moment.

kirillt commented on app/src/main/java/space/taran/arknavigator/mvp/model/repo/scores/PlainScoreStorage.kt:
```diff
@@ -0,0 +1,173 @@
+        lines.add("$STORAGE_VERSION_PREFIX$STORAGE_VERSION")
+
+        val entries = scoreById.filterValues {
+            it > 0 || it < 0
```
`it != 0`

ShubertMunthali replied:
Oh yeah:)

ShubertMunthali replied:
I have it resolved

ShubertMunthali commented on app/src/main/res/layout/fragment_gallery.xml:
```diff
@@ -98,6 +100,42 @@
+            app:layout_constraintEnd_toEndOf="@+id/remove_resource_fab"
+            app:layout_constraintStart_toStartOf="@+id/remove_resource_fab"
+            app:layout_constraintTop_toBottomOf="@+id/increaseScore"
+            app:srcCompat="@drawable/ic_star" />
```
Oh! Yeah! I forgot that one :). I will resolve it
